File: services/go/ride-matching-worker/config_validation.go
package main

import (
	"fmt"
	"os"
	"strings"
)

// requiredEnvironmentVariables enumerates every environment variable the
// worker cannot operate correctly without. The static manifest contract
// check (scripts/testing/check-config-contract.py) reads this table to
// cross-check the Kubernetes manifests against the code.
var requiredEnvironmentVariables = []string{
	"DATABASE_URL",
	"REDIS_URL",
	"INTERNAL_SERVICE_TOKEN",
}

// validateBootConfiguration fails fast at startup, naming every missing or
// invalid required environment variable, instead of crash-looping on the
// first use of a half-configured process.
func validateBootConfiguration() error {
	problems := []string{}
	for _, name := range requiredEnvironmentVariables {
		if strings.TrimSpace(os.Getenv(name)) == "" {
			problems = append(problems, fmt.Sprintf("%s is required", name))
		}
	}
	if token := strings.TrimSpace(os.Getenv("INTERNAL_SERVICE_TOKEN")); token != "" && len(token) < 32 {
		problems = append(problems, "INTERNAL_SERVICE_TOKEN must be at least 32 characters")
	}
	if len(problems) > 0 {
		return fmt.Errorf("invalid boot configuration: %s", strings.Join(problems, "; "))
	}
	return nil
}

File: services/python/market-economics/config_validation.py
"""Boot-time configuration validation for the market-economics service.

Fails fast at startup, naming every missing required environment variable,
instead of crash-looping one variable at a time or silently degrading on
first use. Mirrors services/python/verification-intelligence.
"""

from __future__ import annotations

import os

REQUIRED_ENV_VARS = (
    "INTERNAL_SERVICE_TOKEN",
)


def missing_required_env_vars() -> list[str]:
    return [name for name in REQUIRED_ENV_VARS if not os.getenv(name, "").strip()]


def validate_boot_configuration() -> None:
    missing = missing_required_env_vars()
    if missing:
        raise RuntimeError(
            "invalid boot configuration: missing required environment variables: "
            + ", ".join(missing)
        )
